File: java/epi/src/main/java/moe/ijnji/epi/BalancedBinaryTree.java
// Description:
//
// Check if a binary tree is height balanced.

package moe.ijnji.epi;

import java.lang.Math;
import moe.ijnji.rlib.BinaryTreeNode;

public class BalancedBinaryTree {

    private static int maxHeight(BinaryTreeNode<Integer> root, int height) {
        if (root == null)return height - 1;
        return Math.max(maxHeight(root.left, height + 1), maxHeight(root.right, height + 1));
    }

    private static int minHeight(BinaryTreeNode<Integer> root, int height) {
        if (root == null) return height - 1;
        return Math.min(minHeight(root.left, height + 1), minHeight(root.right, height + 1));
    }

    public static boolean isBalanced(BinaryTreeNode<Integer> root) {
        if (root == null) return true;
        int maxh = Math.max(maxHeight(root.left, 1), maxHeight(root.right, 1));
        int minh = Math.min(minHeight(root.left, 1), minHeight(root.right, 1));
        if (Math.abs(maxh - minh) > 1) return false;
        else return true;
    }

}

File: py/epi/balanced_binary_tree.py
#!/usr/bin/env python3
#
# Description:
#
# Check if a binary tree is height balanced, defined as the difference between
# the shortest branch and the longest branch is not longer than 1.

# from rlib.binary_tree_node import BinaryTreeNode

def check_balance(root):
    def check_recurse(root):
        if root is None: return (True, 0)
        lb = check_recurse(root.left)
        if lb[0] is False: return (False, 0)
        rb = check_recurse(root.right)
        if rb[0] is False: return (False, 0)
        h = max(lb[1], rb[1]) + 1
        return (False if abs(lb[1] - rb[1]) > 1 else True, h)
    return check_recurse(root)[0]
